Readme clarifications (#44)
AbhinavOmprakash/snitch: README.md

> Snitch is inline-defs on steroids

Or:

> Highly enhanced ergonomics for repl-driven development.

Snitch is how I debug and understand data flowing through my system. I use it as my sole debugger (in rare cases I reach for print statements).

> It's not a replacement for a full-fledged debugger, but it's pretty close and will be useful in 90% of the cases. However, unlike a debugger Snitch runs while your application us running, which is both better and worse than a traditional debugger. Better because a debugger freezes your application while you inspect the state. Better because the Clojure REPL will let you modify the code you are debugging, while it is running. Most debugger do not let you do that. Worse because Snitch will only show you the latest value from a series of invocation, where a debugger will let you step to the invocation you are interested in. (BTW. There are ways to collect state from all invocations, even if Snitch doesn't help with it.)

Reviewer: ```suggestion
It's not a replacement for a full-fledged debugger, but it's pretty close and will be useful in 90% of the cases. However, unlike a debugger, Snitch runs while your application is running, which is both better and worse than a traditional debugger. Better because a debugger freezes your application while you inspect the state. Better because the Clojure REPL will let you modify the code you are debugging, while it is running. Most debuggers do not let you do that. Worse because Snitch will only show you the latest value from a series of invocations, whereas a debugger will let you step to the invocation you are interested in. (BTW. There are ways to collect state from all invocations, even if Snitch doesn't help with it.)
```

# Features

> - `defn*`, `*let`, `*fn`, and `defmethod*` macros automating inline `def`s for function arguments and bindings

Reviewer: ```suggestion
- `defn*`, `*let`, `*fn`, and `defmethod*` macros that inject inline `def`s for function arguments and let bindings.
```

- Gives access to the return value of called functions
- Lets you modify and then rerun the function with the snitched invokation arguments
- Support for Clojurescript.
- Editor agnostic (use it along with [CIDER](https://cider.mx), [Conjure](https://conjure.oli.me.uk/), [Calva](https://calva.io), [Cursive](https://cursive-ide.com/), or any REPL enhanced [Clojure editor](https://clojure.org/guides/editors!).

# What people have to say ❤️
> Very handy with those variants of the regular macros. Just add a `*` and you are inline def-ing the args! - Peter Strömberg (co-author of Calva)

# Talk
I gave a [talk](https://www.youtube.com/watch?v=WqilQulsJQc) about snitch at clojure Asia you can watch it
here https://www.youtube.com/watch?v=WqilQulsJQc.

# Usage
## Installation

> I recommend adding snitch to your `~/.lein/profiles.clj` or `deps.edn`.

Reviewer: I don't personally use deps.edn so I don't know much about it, is there some version of profiles.clj for deps? maybe I can add that here.

Author: The version of project.clj for deps is called `deps.edn` 😄. Actually I don't think you need to show either. Linking to the Clojars page lists the coordinates for both.

Reviewer: No, what I meant was that `~/.lein/profiles.clj` adds it to your personal dependencies without adding it to a project's dependencies, is there something for deps.edn that is like that?

An example file would be
```clojure
; Leiningen profiles.clj
{:user {:dependencies [[org.clojars.abhinav/snitch "0.1.16"]]}}

{:dev {:dependencies [[org.clojars.abhinav/snitch "0.1.16"]]}}
```

Once your repl is running you can evaluate:

```clojure
(require '[snitch.core :refer [defn* defmethod* *fn *let]])
```

> in the Clojure REPL and the macros will be interned inside clojure.core & cljs.core, so you don't have to import them in every namespace. It will also make the macros available for your ClojureScript namespaces if your ClojureScript REPL is cloned/spawned off of the Clojure REPL.

Reviewer: > It will also make the macros available for your ClojureScript namespaces if your ClojureScript REPL is cloned/spawned off of the Clojure REPL.

wait, really? that is so cool, I didn't know this!

Author: It's indeed very cool. I add the dependency and do the require in one evaluation and then Snitch is there wherever I need it. And I need it everywhere!

Reviewer: I didn't even know that was possible so thanks for mentioning it 🤯

## Overview
There are four macros `defn*`, `defmethod*` `*fn`, `*let`.

> They are drop-in replacements for `defn`, `defmethod`, `fn`, and `let`, respectively.

Reviewer: Ah, thanks!

These macros create inline defs of the parameters passed to the functions,
and also inside the let bindings of the functions.
This makes it very ergonomic for repl-driven development.
Calling functions defined using the `defn*`, and `*fn` macros will also define symbols that:

> 1. give access to the return value of the functions. `(defn* foo [x y z] ...)` will define `foo<` which holds the return value of calling `foo`.

Reviewer: ```suggestion
1. Give access to the return value of the functions. `(defn* foo [x y z] ...)` will define `foo<` which holds the return value of calling `foo`.
```

> 2. reconstructs the function call. Defining `(defn* foo [x y z] ...)` and then calling `(foo 1 2 3)` will let you reconstruct the call by evaluating `foo>`, you can then evaluate this reconstructed list to repeat the function call. This may not seem like a big deal, but then remember that `foo` may be called indirectly from a chain of function calls, so being able to update the `foo` function and re-call it with the last arguments it was called with is quite convenient.

Reviewer: These are so much clearer than what I came up with! thanks.

Author: You explained it very well in your Clojure Asia presentation. This is basically just that. 😄

> 2. reconstructs the function call. Defining `(defn* foo [x y z] ...)` and then calling `(foo 1 2 3)` will let you reconstruct the call by evaluating `foo>`, you can then evaluate this reconstructed list to repeat the function call. This may not seem like a big deal, but then remember that `foo` may be called indirectly from a chain of function calls, so being able to update the `foo` function and re-call it with the last arguments it was called with is quite convenient.

Reviewer: ```suggestion
2. Reconstructs the function call. Defining `(defn* foo [x y z] ...)` and then calling `(foo 1 2 3)` will let you reconstruct the call by evaluating `foo>`, you can then evaluate this reconstructed list to repeat the function call. This may not seem like a big deal, but then remember that `foo` may be called indirectly from a chain of function calls, so being able to update the `foo` function and re-call it with the last arguments it was called with is quite convenient.
```
```suggestion
2. reconstructs the function call. Defining `(defn* foo [x y z] ...)` and then calling `(foo 1 2 3)` will let you reconstruct the call by evaluating `foo>`, you can then evaluate this reconstructed list to repeat the function call. This may not seem like a big deal, but then remember that `foo` may be called indirectly from a chain of function calls, so being able to update the `foo` function and re-call it with the last arguments it was called with is quite convenient.
```

# Clojurescript support

Snitch works with clojurescript as well. As noted above, if your ClojureScript REPL is spawned/cloned off of the Clojure REPL, you can start your session by requiring the macros in the Clojure REPL, and this will make them available to ClojureScript. If you don't have easy access to the Clojure REPL, you need to include:

```clojure
(:require-macros [snitch.core :refer [defn* defmethod* *fn *let]])
```

in the `ns` form of a namespace.

(That said, usually there is some way to inject code in the Clojure REPL.)

> # Integrating snitch in your workflow.

Reviewer: Thanks for the Re-ogranization of headings and sections

if you have created any specific macros/workflows for your editor of choice, you can open a PR and add it to the readme :)

## Calva

> 

Reviewer: ```suggestion
Here's a [demo](https://www.youtube.com/watch?v=jb1BcYpyOAs) showing how calva and snitch work together by the author of Calva [Peter Strömberg](https://github.com/PEZ) himself :)
```

      "repl": "clj",
      "snippet": "(require '[clojure.repl.deps :refer [add-libs]])\n\n(add-libs '{ org.clojars.abhinav/snitch {:mvn/version \"0.1.16\"}})\n\n(require '[snitch.core :refer [defn* defmethod* *fn *let]])"
    },
    ...
  ],
```

Then press <kbd>cmd+enter</kbd> to get a quick-pick menu which will have the **qol: Add Snitch dependency** option. (If you're not on a mac, use some other key, e.g. <kbd>ctrl+alt+s</kbd> <kbd>enter</kbd>.)

For instrumenting a function definition to use `defn*` instead of `defn` you can define a keyboard shortcut like so:

```jsonc
  {
    // Instrument as snitched defn*
    "key": "cmd+enter",
    "when": "editorLangId == 'clojure' && calva:connected",
    "command": "calva.runCustomREPLCommand",
    "args": {
      "snippet": "${top-level-form|replace|^\\(defn-?|(defn*}"
    }
  },
```

This will evaluate the funcion using `defn*` without editing the file. NB: If you are using some hot-reload tool like [shadow-cljs](https://github.com/thheller/shadow-cljs), saving the file will re-evaluate the function as it is written in the file, removing the snitching. This may or may not be what you want to happen. (Control that save-reflex! 😀)

> You can also bind keys for checking the results of snitched functions and reconstruction the calls to snitched functions:

Reviewer: ```suggestion
You can also bind keys for checking the results of snitched functions and reconstruct the calls to snitched functions:
```

  },
```

> Here's a demo of using these bindings in a shadow-cljs project: https://youtu.be/jb1BcYpyOAs

Reviewer: I added the link to this here so I think we can remove it.
https://github.com/AbhinavOmprakash/snitch/blob/7f49580b292eb67be62890b65524030cbee4fe15/README.md?plain=1#L252-L253
```suggestion
```
